### ISSUE TITLE
Subtract operation fails

### ISSUE THREAD

**Issue by iconexperience** [paperjs/paper.js#890]
Here is another bug where the subtraction fails. An error message is generated and the result of the operation is not correct. This bug is very similar to #889 , but it is not fixed by d89995a

```javascript
var p1 = new CompoundPath();
p1.addChild(new Path({segments:[
    [676, 396.8],
    [445.78, 396.8],
    [426, 260]
], closed:true}));
p1.addChild(new Path({segments:[
    [672, 550.4, 0, 0, 31.74000000000001, 0],
    [633.5999999999999, 732.8, 0, 26.519999999999982]
], closed:true}));
var p2 = new Path({segments:[
    [732, 418, 0, 0, 0, 52.45802606694065],
    [676.3444534355956, 552.3444534355972, 34.385686865893945, -34.38568686589349],
    [630, 704],
    [400, 704],
    [542, 228]
], closed:true});
var res = p1.subtract(p2);
p1.strokeColor = "red";
p2.strokeColor = "green";
res.fullySelected = true;
```
<a href="http://sketch.paperjs.org/#S/lVLLboMwEPwVi1MiuZbxC0PUU36gUo6BA8FuEoVgZEyrKsq/d6FpSpu0Ute+MDPs7uz6FDXl0UZZtDrYUO0iHFXODN8vpUdtjB5RY1/R0h1b1zfmqQy72XyR+7xpY1Ias9ztazMbJCN16uz2aJvQZetBgyDWKlEY8VQRXeArKIQkib6DMxAzRYsBKTCqatdZkwXf2/P834UZRlJSIjCi4+UxSQT9ihjQSXHFOZHpJDBKOCN6/JcpIuMJp9kfPY7DY5fh/dogJMdIxPqzPbiSESE1hQkolQqq5LS9RBEuYHJccClTqQZ3bAoNhrkgXEul4Uid8lRIjB5+gCL95hrqJlRMt0BvICkgN2P6rumrZ287MA0b6vpN8GUVZi27bq0L3h3s0tXOgyiPvDV59EGyG3LrrW0uNGQlz31dv61sbatgDSiGwgt4rRtvy0Pr9jDZKFsX53c=">Here is the Sketch</a>

**Comment by iconexperience** [comment]
I have the feeling that we are doing things too complicated when calculating the winding for these special cases.

This is how the relevant part of the code currently looks like:

```javascript
var x = Curve.getPoint(values, t).x,
    slope = Curve.getTangent(values, t).y,
    counted = false;
if (Numerical.isZero(slope)
    && !Curve.isStraight(values)
    || t < tMin && prevCurve
        && slope * Curve.getTangent(
            prevCurve.values, 1).y <= 0
    // Does the slope over the end change?
    || t > tMax && nextCurve
        && slope * Curve.getTangent(
            nextCurve.values, 0).y <= 0) {
```

The code is supposed to handle 'tips' of a path, which means the horizontal ray (that is cast to determine its crossings with the path and subsequently the winding) merely touches the path. Here are some examples how these cases may look like:

![image](https://cloud.githubusercontent.com/assets/2122773/12165514/e9a5e412-b51e-11e5-8a53-d914d6748ea9.png)

Right now, if the ray touches at the end or beginning of a curve, we are calculating the slope at the touching point and the slope at the equivalent point on the next or previous curve. I think this is too much work, because we should not forget that all curves at this point are monotonically decreasing or increasing in y-direction. This again means that we can only have a tip, if the winding of one curve is the opposite of the winding of the adjacent (next or previous) curve. Therefore I suggest we simplify the code to this:

```javascript
var x = Curve.getPoint(values, t).x,
    slope = Curve.getTangent(values, t).y,
    counted = false;
if (Numerical.isZero(slope)
    && !Curve.isStraight(values)
    || t < tMin && prevCurve
        && winding * prevCurve.winding < 0
    || t > tMax && nextCurve
        && winding * nextCurve.winding < 0) {
```

As you can see I even used the `<`instead of the `<=` comparison again, which seems to work for both our edge cases (this one and #889).

**Comment by lehni** [comment]
Very good! I added this code about 2 years ago without understanding the winding calculations by @hkrish. I think you have a much better understanding of it all by now. I am going to do some tests with this then.

**Comment by lehni** [comment]
Nothing new seems to fail due to this, perfect!

**Comment by iconexperience** [comment]
Actually I had this code change in mind since I worked on `getWinding()`, but somehow I was relucant to propose it, because I was not sure if I understood everything correctly. But as you can see in

https://github.com/paperjs/paper.js/blob/d89995a78199e26b42ca1f1852b5e5a203c5bcac/src/path/PathItem.Boolean.js#L441

we are already doing the same for horizontal curves (`winding === 0`).

**Comment by lehni** [comment]
Ah yes! And it makes sense of course. Perfect!

**Comment by lehni** [comment]
So we're back down to 0 known issues in 20 minutes. Lovin' it.

**Comment by lehni** [comment]
@iconexperience this is the code optimization I talked about before in `getWinding()`: 2fc7684efb3a68ec091f8127448520dce8a43bdd

**Comment by iconexperience** [comment]
That's nice. I was not too happy about my implementation, although it was working well. Should be more clear now and saves some unnecessary loops.

BTW, I am not sure if <a href="https://github.com/paperjs/paper.js/blob/2fc7684efb3a68ec091f8127448520dce8a43bdd/src/path/PathItem.Boolean.js#L410">PathItem.Boolean.js lines 410f</a> are correct or even required, but creating a test case for that would take me probably two days. I will do that if I have a lot of time.

**Comment by lehni** [comment]
I think it was required when we didn't use mono curves, and the ray could touch the middle of a curve that changed direction... But that can't happen anymore now, right?

**Comment by iconexperience** [comment]
with mono curves that should not happen anymore. And if the curves are absolutely horizontal (y values of end points and handles are identical), we have a winding of zero, wich is handled separately. But of course those "almost" horizontal curves are a bit shaky, because imprecision is relatively high.

**Comment by lehni** [comment]
You know it's funny, I looked at that same line earlier, thinking it's the only place that still looks at tangents, and had a hunch that this wasn't required anymore. I removed it locally and can't find any new issues in any of the tests.  I think you are right!

**Comment by iconexperience** [comment]
So remove the lines? Let me test a few minutes.

**Comment by iconexperience** [comment]
@lehni I have found something, but this does not work with the old code base and it does not matter if we remove the mentioned two lines or keep them. But those special zero slop curves are involved.

<a href="http://sketch.paperjs.org/#S/jVJLboMwFLyK5VUiUcvYfjam6ioXqJQlZEGwm0QhODLQqopy99rk06CoUi02b2bwvBk44bY6WJzj5d729RYnuHYmzp+VR8eq36boDbX2Cy3c4eiG1rwHbDZ/LX3ZjjSpjFlsd42ZRdXInjq7Odi27/KiACUTxLUk2SpBBRdAVPYLoIIJTYKCCU4YfThsFei6cZ01ee8He57/31MyQTKRIABKhIy+UrHrGCfFRk9Jsz9MbtnZNfuTA2ecTLYNCVg2gdIE0fGBlGjIlGY65SCBQkDjFiKlRD+cUIuiYnpt1L2wdAoyFZV8+vJNKYMPpUoJBlxBrBokAeAQNtJCA3vOfI/sbRcCXxruhnXvq7qfjT1cNIEnH0PTfC9tY+vemqCOVzx8mK73bm8XrnE+kCX21pT4zrMnfuOtba+K8O+tva32R7cLLeO8WJ1/AA==">Here is the Sketch (fails, too)</a>

```javascript
var path1 = new CompoundPath();
path1.addChild(new Path({segments:[[576, 396.8], [345.78, 396.8],  [249.6, 243.20000000000002]], closed:true}));
path1.addChild(new Path({segments:[[624.84, 550.46], [672, 550.4], [729.6, 608]], closed:true}));
var path2 = new Path({segments:[[323.0000000000002, 228.0000000000001, 0, 0, 51.95879291356505, 0], [410.9999999999998, 704.0000000000002], [-21.000000000000227, 703.9999999999998], [-21.692900774253758, 356.5535001949528]], closed:true});
var res = path1.subtract(path2);
res.fullySelected = true;
path1.strokeColor = "red";
path2.strokeColor = "green";
```

This is really an edge case of an edge case. I have rotated my test very slightly to create those "almost horizontal" lines, and even then it took me a while to get the error.Oh, and the result looks corect, so we could just ignore it. But of course we won't :)

**Comment by lehni** [comment]
Hahahah. Damn you!

**Comment by lehni** [comment]
Funny, even this fixes it: `path2.rotate(0.0000000001);`

**Comment by iconexperience** [comment]
It's the points P1 and P2 that are causing trouble.

![image](https://cloud.githubusercontent.com/assets/2122773/12171897/a595dfc8-b54d-11e5-93e6-3fc02333966e.png)

**Comment by iconexperience** [comment]
And these are the coordinate of the points:

```
var p1 = new Point(373.9911696902853, 550.4000000000001);
var p2 = new Point(373.9911696902853, 704.0000000000001);
```

**Comment by lehni** [comment]
It took me y while but I've isolated it to this simple situation returning a wrong winding value:

[Sketch](http://sketch.paperjs.org/#S/TVBBboMwEPzKlgsgIQdIcIKjnvKBSjn0QFBF7FViQezKNu0hyt9rA2m7kqWdWe3MeO+R6m4YsejYo+PXKIu4FgF/dQY+O3ct4BUUfsOb75O7xcsNlbOsaWi5IbtNBlWVkw1tM2jotlxgQNuyJjQDmu9aD/mgLQrmzIiPdH8yJzWJE+uM7vGgB228UWxQxNN0stdSuad96JP1dk3quihoTeu83FXF4pf/VTGrPyOTgzR8wOQeSPDFfXw0bBbPnqzphBwtg/Uv8y8Ym3OFyZI9PK6V1QOSQV+S+S8fF3TvUgmpPBPk03QPqxUcr3ocBJwRcrBScQTpYgt6dFYKfAli/u5ng10/rdmINe3jBw==)

![screen shot 2016-01-07 at 14 59 35](https://cloud.githubusercontent.com/assets/213275/12172129/4ad5bcd2-b54f-11e5-8786-3f0285aea6fa.png)

**Comment by lehni** [comment]
Haha, same coordinates : )

**Comment by lehni** [comment]
And here the situation of the 2nd coordinate that you found, and I wasn't aware of yet:

[Sketch](http://sketch.paperjs.org/#S/ZVLLboMwEPyVLZeARAm2MWCinvIDlXLogUQVgVVihdiVTdpDlH+vTUQa1JUtrWdnZx/yNVDNGYMq2JxwaI9BHLS68+/vxsBXMxwpvIHCH3h3fni1eDijGmxV14yyJH0yGgOl5QwiMaTj4SQRvCwEFYTxnKfcobsY6oykiXiyMoYizeaynvdKyRykhWeyefLEzF2dNC2KjHJWcKfpaiacM+46EpngtNw5attri101mAveotXWbNU4bmIHo0+41r02bvSFwW4xRv0dl6KlGqaleD9khW+EkFzkIqUlZ/+nINFDY1pmspam7TG8ehCctW6xaKp7gXhCTdPJi62APZCnBqt7fz5y+6vQamV1j0mvD+F9ps8DDh9SdVI5xMtH0QqWS9gc9aXvYI9AwErVIshhYUEqKzt88VruQ+wNNqcxywZVvbv9Ag==)

**Comment by lehni** [comment]
Do you want to look into what's causing these two, @iconexperience?

**Comment by iconexperience** [comment]
Yes, I will. The first one is strange, I thought these cases should have been fixed.

**Comment by lehni** [comment]
The horizontal lines in either situation aren't fully horizontal, are they... Do we need more tolerance, or less, when deciding for the horizontal treatment?

**Comment by lehni** [comment]
Oh wait, that code isn't even in use when calling _getWinding()...

**Comment by lehni** [comment]
Oh but this is interesting! Passing `true` to the calls of `_getWinding()` in both examples leads the correct result!

**Comment by iconexperience** [comment]
Yes, but how do we know if we encounter horizontal lines?

**Comment by lehni** [comment]
Well the only place where this is actually calculated is outside of `getWinding()`, when calling it from `propagateWinding()`... But what we really need is code inside `getWinding()` that figures this out too. Because in other situations, e.g. when calling `_getWinding()` from `contains()`, it doesn't provide a proper value for `horizontal`and the above examples fail there too (not the first one now because of the added bounds check, but the 2nd one)

**Comment by lehni** [comment]
And the situation that comes to bite us here internally is when `getWinding()` calls itself with new coordinates, which happen to have a ray going onto a horizontal curve.

I'm not sure what a proper solution will consist of. Seems like quite a tricky one to solve nicely, actually!

**Comment by lehni** [comment]
Perhaps the solution is in the code that currently handles this *external* `horizontal` parameter... But instead of branching in the beginning of `getWinding()`, we have to switch to this strategy as soon as we realize the ray is horizontally touching such a curve in any of the monoCurves?

The strategy consists of shooting a vertical ray from the current point in both directions until meeting the next curve, and then taking the average of the current y-coordinate and the closest encountered point on another curve (guaranteed to be inside the same shape as the current point), and carry on sampling with that location.

**Comment by lehni** [comment]
Hmm but I think that's only concerning the 2nd case here, where the line is actually almost horizontal (the slope is `1.052655904829778e-15`) so should be considered horizontal, but doesn't receive a winding of `0` yet because of the strict comparison here: https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L938

I can look into that case, if you'd like to tackle the first one? There something else is going down...

**Comment by lehni** [comment]
Remember when I said we probably need comparisons with epsilon here? https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L434
Well the 2nd case here is such a situation now.

**Comment by iconexperience** [comment]
I am trying to find a solution for the first case by simply adjusting the algorithm. Will take a few minutes to come to a conclusion.

**Comment by lehni** [comment]
Sounds good! Meanwhile I can confirm that adding epsilons to the horizontal checks in `_getMonoCurves()` and `_getWinding()` solves the second case and doesn't produce new ones in any of the available tests. The only question I'm not sure about yet is what epsilon use in [this comparison](https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L434).

**Comment by iconexperience** [comment]
Regarding the first case I think we are doing it wrong at the moment and maybe too complicated.

Currently we check if an intersection at the end of one curve is also at the beginning of the next curve. If so, we make sure that the intersection is counted only once. But IMHO this should only be done if winding does not change direction between curves. If it changes direction, let's just count both intersections, in and out. Let's see if I can turn this into code.

**Comment by lehni** [comment]
That sounds reasonable, as long as we alway catch both intersections in such situations (which we should)

**Comment by lehni** [comment]
I just made another interesting discovery: It looks like the sampling in three points to get the winding contribution of a chain of segments in `propagateWinding()` might not be required anymore. I switched to sampling only one location in the center, and there are no failing cases in the unit tests, and only one failing case in your test suite (1419). The code executes much faster, so I believe it's well worth attempting to find the source of that one failing case.

**Comment by iconexperience** [comment]
That sounds great. My attempts very quite frustrating until the great Eureka a few minutes ago.

I found that we can simplify `getWinding()` a lot. Just look at this (the non-horizontal part):

```javascript
    function getWinding(point, curves, horizontal) {
		var epsilon = 2e-7,
			px = point.x,
			py = point.y,
			windLeft = 0,
			windRight = 0,
			length = curves.length,
			roots = [],
			abs = Math.abs;
		if (horizontal) {
			var yTop = -Infinity,
				yBottom = Infinity,
				yBefore = py - epsilon,
				yAfter = py + epsilon;
			for (var i = 0; i < length; i++) {
				var values = curves[i].values,
					count = Curve.solveCubic(values, 0, px, roots, 0, 1);
				for (var j = count - 1; j >= 0; j--) {
					var y = Curve.getPoint(values, roots[j]).y;
					if (y < yBefore && y > yTop) {
						yTop = y;
					} else if (y > yAfter && y < yBottom) {
						yBottom = y;
					}
				}
			}
			yTop = (yTop + py) / 2;
			yBottom = (yBottom + py) / 2;
			if (yTop > -Infinity)
				windLeft = getWinding(new Point(px, yTop), curves);
			if (yBottom < Infinity)
				windRight = getWinding(new Point(px, yBottom), curves);
		} else {
			var xBefore = px - epsilon,
				xAfter = px + epsilon,
				start,
				end = 0;
			while (end < length) {
				start = end;
				var lastCurve;
				for (var i = start; i < length; i++) {
					var curve = curves[i];
					if (curve.winding) {
						lastCurve = curve;
					}
					if (curve.next !== curves[i + 1]) {
						end = i + 1;
						break;
					}
				}
				var prevWinding = lastCurve.winding,
					curve = null;
				for (var i = start; i < end; i++) {
					curve = curves[i];
					var values = curve.values,
						winding = curve.winding;
					if (winding && ((py >= values[1] && py <= values[7])
						|| (py >= values[7] && py <= values[1]))) {
						if ((py != values[1] || winding * prevWinding < 0) && Curve.solveCubic(values, 1, py, roots, 0, 1) === 1) {
							var t = roots[0];
							var x = Curve.getPoint(values, t).x;
							if (x <= xBefore) {
								windLeft += winding;
							} else if (x >= xAfter) {
								windRight += winding;
							}
						}
						prevWinding = winding;
					}
				}
			}
		}
		return Math.max(abs(windLeft), abs(windRight));
	}
```

No `tMin` or `tMax`! The reason is this: because we only have mono curves, and we are using no tolerances in y direction, our ray can only cross two adjacent curves under these conditions:
- py is exactly at the end point of the curve (=start point of the next curve)
- the curves have opposite windings
If the curves have opposite windings, we can count both crossings. The other case can be detected by checking for `py === values[1]` and `winding === prevWinding` (where `prevWinding` is the winding of the previous non-horziontal curve.

This seems to work with my test suite and solves the second bug of this issue (while the first still fails). It will probably require a little more work, but isn't this a nice discovery?

**Comment by lehni** [comment]
Wowzie!!! That is so very very very cool! I will put it through tests right now.

Meanwhile I found out more about my discovery above as well, but will write about that in a bit.

Very exciting, all this :+1:

**Comment by iconexperience** [comment]
Nervously waiting for your results... :watch: :)

**Comment by lehni** [comment]
Did you run the internal unit-tests? I currently have 12 failed tests there after using your version of `getWinding()` One is #890, the others are hand-crafted tests for Path#contains() (the assumptions for might be wrong...)

**Comment by iconexperience** [comment]
I do not handle the case when a point in on a horizontal curve, yet. But actually, that may even more simplify the code. I will be right back with results.

**Comment by iconexperience** [comment]
I got the failed tests down to 6 with handling horizontal curves. It will take a bit more work (touch points don't seem to be handles correctly, yet), but I think we will get there and the result will be smaller and faster than our current code.

**Comment by lehni** [comment]
Cool! There is no rush with any of this, so take your time : )

**Comment by iconexperience** [comment]
Fails down to these two:

![image](https://cloud.githubusercontent.com/assets/2122773/12181300/b31b581a-b581-11e5-889d-cbf696d32597.png)

**Comment by lehni** [comment]
Ha, what is that even! Could be inside, could be outside : )

**Comment by iconexperience** [comment]
Left one inside, right one outside in my opinion.

**Comment by lehni** [comment]
It depends on what that line is doing, I guess? But you really could argue that both of them should be inside... Is this part of the unit tests?

**Comment by iconexperience** [comment]
Yes. This was one of the reasons why it took me so long to solve #819. I would simply ignore the horizontal lines in this case, that's why I think the left one is inside and the right one outside.

**Comment by lehni** [comment]
Right! that makes sense, yes.

**Comment by iconexperience** [comment]
Here is my result. I am totally exhausted. Comments will follow tomorrow, but if you have a few minutes, maybe you can run some tests. And it can certainly be made a bit more elegant.

```javascript
   function getWinding(point, curves, horizontal) {
		var epsilon = 2e-7,
			px = point.x,
			py = point.y,
			windLeft = 0,
			windRight = 0,
			length = curves.length,
			roots = [],
			abs = Math.abs;
		if (horizontal) {
			var yTop = -Infinity,
				yBottom = Infinity,
				yBefore = py - epsilon,
				yAfter = py + epsilon;
			for (var i = 0; i < length; i++) {
				var values = curves[i].values,
					count = Curve.solveCubic(values, 0, px, roots, 0, 1);
				for (var j = count - 1; j >= 0; j--) {
					var y = Curve.getPoint(values, roots[j]).y;
					if (y < yBefore && y > yTop) {
						yTop = y;
					} else if (y > yAfter && y < yBottom) {
						yBottom = y;
					}
				}
			}
			yTop = (yTop + py) / 2;
			yBottom = (yBottom + py) / 2;
			if (yTop > -Infinity)
				windLeft = getWinding(new Point(px, yTop), curves);
			if (yBottom < Infinity)
				windRight = getWinding(new Point(px, yBottom), curves);
		} else {
			var xBefore = px - epsilon,
				xAfter = px + epsilon,
				start,
				end = 0;
			while (end < length) {
				start = end;
				var firstCurve = null,
					lastCurve;
				for (var i = start; i < length; i++) {
					var curve = curves[i];
					if (curve.winding) {
						if (!firstCurve)
							firstCurve = curve;
						lastCurve = curve;
					}
					if (curve.next !== curves[i + 1]) {
						end = i + 1;
						break;
					}
				}
				var prevWindingVals = lastCurve.values,
					prevWinding = lastCurve.winding,
					curve = null;
				for (var i = start; i < end; i++) {
					curve = curves[i];
					var values = curve.values,
						winding = curve.winding;
					if (winding && ((py >= values[1] && py <= values[7])
						|| (py >= values[7] && py <= values[1]))) {
						if (py === values[1]) {
							var x = values[0];
							if (winding * prevWinding < 0) {
								if (x > xBefore && x < xAfter) {
									++windLeft;
									++windRight;
								} else if (x <= xBefore) {
									windLeft += winding;
								} else if (x >= xAfter) {
									windRight += winding;
								}
							} else if (x <= xBefore && prevWindingVals[6] >= xBefore) {
								windLeft += winding;
							} else if (x >= xAfter && prevWindingVals[6] <= xAfter) {
								windRight += winding;
							}
						} else if (Curve.solveCubic(values, 1, py, roots, 0, 1) === 1) {
							var t = roots[0];
							var x = Curve.getPoint(values, t).x;
							if (x <= xBefore) {
								windLeft += winding;
							} else if (x >= xAfter) {
								windRight += winding;
							}
						}
						prevWinding = winding;
						prevWindingVals = values;
					}
				}
			}
		}
		return Math.max(abs(windLeft), abs(windRight));
	}
```

**Comment by iconexperience** [comment]
I just noticed that maybe `prevWinding`should be set to zero or null if we encounter a non-horizontal that is completely above or below `py`.

**Comment by iconexperience** [comment]
Here everything seems to work, only the infamous glitch 244 is back.

**Comment by lehni** [comment]
I'll have a look now. But it's time to rest!

**Comment by lehni** [comment]
That's strange you're still seeing issues with glitch 244! I have zero glitches here, it all works perfectly!

**Comment by iconexperience** [comment]
Maybe my famous code experiments. But doesn't it work really well, even on both #890 issues?

**Comment by lehni** [comment]
Yes most def! Shall I brush it up a little and merge it in already? You can write your comments tomorrow then?

**Comment by iconexperience** [comment]
Yes please. For me it's bedtime!

**Comment by lehni** [comment]
I am letting @hkrish's unit test run through it too. That's another 80k of crazy tests or so, compared against Java2D boolean ops : )

**Comment by lehni** [comment]
Man, I can't believe how much simpler this code has become. Really impressive work, @iconexperience!

**Comment by lehni** [comment]
So here we go: eba9601601590c3645fc4dca10c27383f8adeabc

I added some comments already but there's one part that will benefit from your writing. 

I also managed to simplify _getMonoCurves() quite a bit as a result of the new simplified code. It's all very cool!

**Comment by iconexperience** [comment]
I have changed the code in #895 again to make it even better understandable. I think this is the best version so far. Maybe you can check if it keeps your coding and formatting standards.

**Comment by lehni** [comment]
Yes it's extremely fascinating, and addictive... But hopefully we can put that part of the code to rest now for a bit! And here something that should put your mind at ease. 120k+ tests, 0 failures.

```
[union] Testing paperjs unite operator.
 
  Test Report         [union]
----------------------------------------------------
  Tests run         - 40576
  Tests passed      - 40576 100.0%
  Tests failed      - 0 0.0%

  Elapsed           - 44 : 40 : 335.9
  Boolean op. time  - 16 : 44 : 196.3
  Boolean avg. time - 0 : 0 : 24.7

[subtraction] Testing paperjs subtract operator.
 
  Test Report         [subtraction]
----------------------------------------------------
  Tests run         - 40576
  Tests passed      - 40576 100.0%
  Tests failed      - 0 0.0%

  Elapsed           - 44 : 19 : 789.6
  Boolean op. time  - 16 : 44 : 163.3
  Boolean avg. time - 0 : 0 : 24.7

[intersection] Testing paperjs intersect operator.
 
  Test Report         [intersection]
----------------------------------------------------
  Tests run         - 40576
  Tests passed      - 40576 100.0%
  Tests failed      - 0 0.0%

  Elapsed           - 44 : 9 : 704.9
  Boolean op. time  - 16 : 34 : 341.4
  Boolean avg. time - 0 : 0 : 24.5
```

**Comment by iconexperience** [comment]
Very, very nice!

The `getWinding()` was the last part of the boolean code that I felt uncomfortable with (https://github.com/paperjs/paper.js/issues/875#issuecomment-167888675), but I like it very much now. Therefore I will take some days off now from paper.js, but I will be listening in case there is something where I can assist.

**Comment by lehni** [comment]
The last part I'm uncomfortable with is `propagateWinding()`... I have a feeling that this is prone to some errors still in edge cases, as the sampling happens in random places. Even if we're sampling in 3 random (or evenly distributed) places, we can still hit the three wrong spots. Since this is performed after all the intersections are divided, and we're only distributing winding between intersections (areas in which the winding contribution should be the same), the only reason for this must be those epsilons in `getWinding()` and proximity to other curves... I thought yesterday that I was able to remove the looping, but I was wrong. This is something we could still improve! But not now : )

Have a good, well deserved rest! I should do the same...
